### ISSUE TITLE
Robustness linters

### ISSUE THREAD

**Issue by jimhester** [jimhester/lintr#48]
Suggestions from @hadley
- [ ] Require drop = FALSE whenever [ is used with two or more args
- [ ] Require na.rm argument for any function that has it (to force you to
  think about missing values during development)
- [x] Don't use of require()/library() inside a function
- [x] Don't change global state: setwd(), options(), par() etc.  Could
  have next package robustr that contained alternatives (i.e. the with_*
  functions from devtools)
- [ ] Always turn partial matching warning options on (not really code
  linting, but related)
- [ ] Avoid functions that use NSE (with(), subset(), transform() etc)
- [ ] Avoid functions that return different types of output (e.g. sapply())
- [x] `TRUE` and `FALSE` instead of `T` and `F`.
- [ ] `return()` calls at the end of functions (result of last statement is always returned, so the explicit `return()` is redundant (and also slower as it requires an additional function call).
- [x] use of `1:length(x)` in for loops (dangerous if length(x) can be 0) as `1:0` returns `c(1, 0)`, should use `seq_len(length(x))` instead.
- [ ] Iterative appending to vectors within loops rather than pre-allocation the vectors and filling. (the former is quadratic in time complexity, as R has to keep re-sizing the vector on every assignment.

``` r
# bad
for (i in seq_len(10)) {
  x <- c(x, i)
}

# good
x <- integer(10)
for (i in seq_len(10)) {
  x[i] <- i
}
```

**Comment by fangly** [comment]
The following items have been implemented recently:
* `1:x` usages are caught by `seq_linter()`
* `T_and_F_symbol_linter` catches "T" and "F" instead of `TRUE` and `FALSE`.
* `require`, `library`, `setwd()`, `options()`, `par()`, `sapply` are linted by `undesirable_function_linter()` using the functions listed by `default_undesirable_functions()`. Additional functions to lint can easily be added.
